### ISSUE TITLE
Implement GP-based Evolution of Optimizers

### ISSUE THREAD

**Issue by andrewni2002** [andrewni420/GPU-PUSH#12]
Much like how a neural network is a program that takes in an input and returns an output, an optimizer is a program that takes in the current training state (containing things like parameter values, momentum, current epoch, etc.) and the gradient and returns the updated training state. 

Although a neural network can be parameterized by its parameters, optimizers are usually only parameterized by a small set of hyperparameters. 

Optimizers can:
- [ ] Create a new pytree of the same shape as the input, initialized to a certain value (Creates a variable slot)
    - [ ] Is it possible to have entries of any other shape besides either being a single value or having the same pytree structure as the parameters?
- [ ] Reference the parameters or the gradient
- [ ] Perform operations on pytrees, floats, and ints:
    - [ ] add, subtract, multiply, divide
    - [ ] max, min (elementwise / aggregated)
    - [ ] sum, mean
    - [ ] square, abs, sqrt, sign
- [ ] Logical operations and comparisons (e.g. for backtracking line search):
    - [ ] if/else
    - [ ] for/while
    - [ ] and/or/not
    - [ ] >, >=, <, <=, ==
    - [ ] Elementwise versions of the above (where relevant)
- [ ] Similar to RNNs, during execution, optimizers can use entries to update variable slots by value or index when shapes are compatible.
At the end of execution, optimizers will return the top pytree-shaped entry as the updated parameters, as well as the (potentially updated) variable slots
